nomenclature/processor/requiredData.py

```diff
@@ -0,0 +1,140 @@
+import logging
+from pathlib import Path
+from typing import List, Optional, Union, Tuple
+
+import yaml
+from pyam import IamDataFrame
+import pydantic
+from pydantic import BaseModel, validator
+from pydantic.error_wrappers import ErrorWrapper
+
+from nomenclature.definition import DataStructureDefinition
+from nomenclature.processor.utils import get_relative_path
+from nomenclature.error.requiredData import RequiredDataMissingError
+
+logger = logging.getLogger(__name__)
+
+
+class RequiredData(BaseModel):
+
+    variable: List[str]
+    region: Optional[List[str]]
+    year: Optional[List[int]]
+    unit: Optional[str]
+
+    @validator("variable", "region", "year", pre=True)
+    def single_input_to_list(cls, v):
+        return v if isinstance(v, list) else [v]
+
+    def validate_with_definition(self, dsd: DataStructureDefinition) -> None:
```

Comment:
This is a duplication of the function in the RegionProcessor, right? I guess you thought about streamlining and had a good reason for not doing it, right?

Reply:
It's less of a duplication but more implementing functionality that's also needed here. The logic is that `requiredData` and as a result `requiredDataValidator` should not require anything that's not covered by the codelists as it would otherwise be impossible to satisfy. E.g. if in `requiredData` we require the variable _Final Energy_ to be present for the region _World_ both must exist in the project's codelists.
`requiredDataValidator` also has `validate_with_definition` which loops over each `requiredData` element and validates it.

Reply:
Maybe move this as a validation-method into the codelist and have the processors pass their lists to that method?

Reply:
Ah sure, that works too. Should I do the same thing for the `RegionProcessor` as well?

Reply:
Yes, I think that would simplify the package... Thanks!

Reply:
Sounds good, I'll do that.